jctools-core/src/main/java/org/jctools/queues/atomic/BaseLinkedAtomicQueue.java

```diff
@@ -53,6 +53,51 @@ protected final void spConsumerNode(LinkedQueueAtomicNode<E> node) {
         throw new UnsupportedOperationException();
     }
 
+    /**
+     * {@inheritDoc}
+     * <p>
+     * This method is only safe to call from the consumer thread!
+     */
+    @Override
+    public final boolean remove(Object o) {
+        if (null == o) {
+            throw new NullPointerException();
+        }
+
+        for (LinkedQueueAtomicNode<E> prevConsumerNode = null, currConsumerNode = getNextConsumerNode(lpConsumerNode());
+             currConsumerNode != null;
+             prevConsumerNode = currConsumerNode, currConsumerNode = getNextConsumerNode(currConsumerNode)) {
+            if (o.equals(currConsumerNode.lpValue())) {
+                LinkedQueueAtomicNode<E> nextNode = getNextConsumerNode(currConsumerNode);
+                if (nextNode != null && prevConsumerNode != null) {
+                    // We are removing an interior node.
+                    prevConsumerNode.soNext(nextNode);
+                } else if (prevConsumerNode == null && nextNode == null) {
+                    // We are removing the only node.
+                    currConsumerNode.spValue(null);
+                    spConsumerNode(currConsumerNode);
+                } else if (nextNode == null) {
+                    // We are removing the last node.
+                    prevConsumerNode.soNext(null);
+
+                    // If the producer node is pointing to the node we are removing then we should try to atomically
+                    // update the reference to move it to the previous node.
+                    if (!producerNode.compareAndSet(currConsumerNode, prevConsumerNode)) {
+                        // If the producer(s) have offered more items we need to remove the currConsumerNode link.
+                        while ((nextNode = currConsumerNode.lvNext()) == null);
+                        currConsumerNode.soNext(currConsumerNode);
+                        prevConsumerNode.soNext(nextNode);
+                    }
+                } else {
+                    // We are removing the first node.
+                    currConsumerNode.spValue(null);
+                    spConsumerNode(currConsumerNode);
+                }
```

Comment:
I consolidated to 3 blocks: interior, consumer, producer. However more simplification is welcome.